### PR TITLE
Initial work on Ice for Python API reference documentation

### PR DESCRIPTION
**Pull request by pepone**
This PR updates Ice for Python API docs:

The documentation is generated with sphinx, one of the standard tools for documenting Python projects.

The PR is getting quite large, I want to get some feedback before I add more changes.

### CODE REVIEW

File: cpp/src/slice2py/PythonUtil.cpp [hunk 1]
```diff
@@ -408,58 +408,58 @@ Slice::Python::CodeVisitor::visitInterfaceDecl(const InterfaceDeclPtr& p)
...
+            _out << nl;
+            _out << nl << "Args:";
+            _out << nl << "  result: The result (or result tuple) of the invocation.";
+            _out << nl << "  current: The Current object passed to the invocation.";
+            _out << nl;
+            _out << nl << "Returns";
+            _out << nl << "  An object containing the marshaled result.";
+            _out << nl << tripleQuotes;
+            _out << nl << "return IcePy.MarshaledResult(result, _M_" << getAbsolute(p) << "._op_" << fixedOpName
+                 << ", current.adapter.getCommunicator().getImpl(), current.encoding)";
```
Comment (pepone) [inline cpp/src/slice2py/PythonUtil.cpp:440]: Need to update this to use numpy style..

File: python/modules/IcePy/Current.cpp [hunk 2]
```diff
@@ -308,12 +308,20 @@ PyObject*
 IcePy::createCurrent(const Ice::Current& current)
 {
     //
-    // Return an instance of IcePy.Current to hold the current information.
+    // Return an instance of Ice.Current to hold the current information.
```
Comment (pepone) [inline python/modules/IcePy/Current.cpp:311]: Ice.Current is an abstract class defined in Python and implemented by the private class CurrentI. The CurrentI class acts as a wrapper around IcePy.Current. This setup, while somewhat involved, is designed to avoid exposing IcePy types directly in our public API.

File: python/modules/IcePy/Endpoint.cpp [hunk 3]
```diff
@@ -134,9 +134,9 @@ namespace IcePy
     PyTypeObject EndpointType = {
         /* The ob_type field must be initialized in the module init function
          * to be portable to Windows without using C++. */
-        PyVarObject_HEAD_INIT(0, 0) "IcePy.Endpoint", /* tp_name */
-        sizeof(EndpointObject),                       /* tp_basicsize */
-        0,                                            /* tp_itemsize */
+        PyVarObject_HEAD_INIT(0, 0) "Ice.Endpoint", /* tp_name */
```
Comment (pepone) [inline python/modules/IcePy/Endpoint.cpp:137]: need to revert this change. I was testing the direct exposure of this component as part of the Ice module and documenting it in a .pyi file.

Using the autoapi extension (https://sphinx-autoapi.readthedocs.io/en/latest/) seemed to work initially. However, I reverted to the autodoc extension because autoapi does not support our current generated code due to the conditional class definitions.

File: python/modules/IcePy/ObjectAdapter.cpp [hunk 4]
```diff
@@ -1768,7 +1768,7 @@ IcePy::wrapObjectAdapter(const Ice::ObjectAdapterPtr& adapter)
     {
         return nullptr;
     }
-    PyObject* wrapperType = lookupType("Ice.ObjectAdapterI.ObjectAdapterI");
+    PyObject* wrapperType = lookupType("Ice._ObjectAdapterI.ObjectAdapterI");
```
Comment (pepone) [inline python/modules/IcePy/ObjectAdapter.cpp:1771]: The `_` prefix is used for private modules. I renamed private classes to use this prefix to help exclude them from the generated documentation.

File: python/modules/IcePy/Proxy.cpp [hunk 5]
```diff
@@ -2249,8 +2249,7 @@ IcePy::createProxy(const Ice::ObjectPrx& proxy, const Ice::CommunicatorPtr& comm
 {
     if (!type)
     {
-        PyTypeObject* proxyType = &ProxyType; // Necessary to prevent GCC's strict-alias warnings.
-        type = reinterpret_cast<PyObject*>(proxyType);
+        type = lookupType("Ice.ObjectPrx");
```
Comment (pepone) [inline python/modules/IcePy/Proxy.cpp:2252]: Ice.ObjectPrx is the public base proxy type, it delegates to the parent class IcePy.ObjectPrx.

File: python/python/Ice/Future.py [hunk 6]
```diff
@@ -183,3 +183,5 @@ def _warn(self, msg):
     StateRunning = "running"
     StateCancelled = "cancelled"
     StateDone = "done"
+
+__all__ = ["Future", "wrap_future", "FutureBase"]
```
Comment (pepone) [inline python/python/Ice/Future.py:187]: The __all__ variable specifies the public API and controls what is exposed when using the import * syntax. This prevents duplication warnings during documentation generation by avoiding the re-export of imported items, which would otherwise result in warnings indicating they are defined in multiple places.

File: python/python/Ice/_CurrentI.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+
+from .Current import Current
```
Comment (pepone) [inline python/python/Ice/_CurrentI.py:2]: The private implementation of Ice.Current.

File: python/test/Ice/custom/ArrayUtil.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) ZeroC, Inc. All rights reserved.
```
Comment (pepone) [inline python/test/Ice/custom/ArrayUtil.py:1]: I copied it here because of Ice._ArrayUtil is not exported anymore.

File: python/modules/IcePy/Current.cpp (hunk 2 above)
Comment (bernardnormier) [reply python/modules/IcePy/Current.cpp:311]: It sounds very complicated for something very simple such as Current.

Why is Current abstract in the first space? We can't just use a concrete Ice.Current fully defined in Python?

File: python/python/Ice/_CommunicatorI.py [hunk 9]
```diff
@@ -1,10 +1,10 @@
...
-from .ObjectAdapterI import ObjectAdapterI
-from .ImplicitContextI import ImplicitContextI
-from .LoggerI import LoggerI
+from ._PropertiesI import PropertiesI
+from ._ObjectAdapterI import ObjectAdapterI
+from ._ImplicitContextI import ImplicitContextI
+from ._LoggerI import LoggerI
 from .Logger import Logger
 
 class CommunicatorI(Communicator):
```
Comment (bernardnormier) [inline python/python/Ice/_CommunicatorI.py:10]: In many languages, we merged Communicator and CommunicatorI, Properties and PropertiesI (and more). Could we do the same in Python?

File: python/modules/IcePy/Current.cpp (hunk 2 above)
Comment (pepone) [reply python/modules/IcePy/Current.cpp:311]: The `IcePy.Current` class holds a pointer to the C++ current object, and its getter lazily creates the Python objects for each field. This optimization avoids creating fields that might never be accessed by the dispatch. However, defining a concrete `Current` class in Python and fully initializing it from the start would be simpler.

File: python/python/Ice/_CommunicatorI.py (hunk 9 above)
Comment (pepone) [reply python/python/Ice/_CommunicatorI.py:10]: We can, the only caveat is that the constructor takes a reference to the corresponding IcePy type. I guess this is fine if we keep it undocumented.

File: python/modules/IcePy/Current.cpp (hunk 2 above)
Comment (bernardnormier) [reply python/modules/IcePy/Current.cpp:311]: Assuming we can reuse the same ObjectAdapter and Connection object over and over (not create new one for each dispatch), I'd create a populated Python Ice.Current, even if the application doesn't use it (which is very common).

Comment (pepone) [reply python/modules/IcePy/Current.cpp:311]: Fixed in #2552

File: python/python/Ice/_CommunicatorI.py (hunk 9 above)
Comment (pepone) [reply python/python/Ice/_CommunicatorI.py]: Fixed in #2554

File: python/python/Ice/ObjectPrx.py [hunk 10]
```diff
@@ -0,0 +1,727 @@
+
+import IcePy
+
+class ObjectPrx(IcePy.ObjectPrx):
```
Comment (bernardnormier) [inline python/python/Ice/ObjectPrx.py:4]: I would prefer to use delegation to the C++ object, like for example Communicator. Am I missing something?

**Review by bernardnormier — COMMENTED** [review]
Looks good.

For LoggerI in file _LoggerI.py, is this naming inconsistency on purpose?

**Review by bernardnormier — APPROVED** [review]

**Comment by pepone** [comment]
> For LoggerI in file _LoggerI.py, is this naming inconsistency on purpose?

The `_` prefix is used by convention in Python for private modules/members.

File: .github/workflows/docs.yml [hunk 11]
```diff
@@ -48,6 +51,17 @@ jobs:
         run: |
           make doc
 
+      - name: Generate Python Documentation
+        working-directory: ./python/docs
+        run: |
+          make -C ../cpp slice2py Ice IceDiscovery IceLocatorDiscovery
```
Comment (pepone) [inline .github/workflows/docs.yml:57]:
```suggestion
          make -C ../../cpp slice2py Ice IceDiscovery IceLocatorDiscovery
```
